### DISCUSSION TITLE
Shared templates folder

### DISCUSSION THREAD

**Discussion by muellerj** [jeremyevans/roda#229] [General]
Given I have two apps, which have some of their templates in common (layout, flash, header, footer, not_found, etc.), is there a way to tell the `render` plugin the following instructions:

Please look for `template_a` in the following `PATHS` and take the the first one you find:

```
PATHS = [ "/srv/app-a/views", "/srv/common/views" ]
```

Ideally, I'd like to write `view "template_a"` and resolve the actual template transparently based on the contents of the two (or more) directories.

I've had a look in https://github.com/jeremyevans/roda/blob/master/lib/roda/plugins/render.rb#L774, but this seems to take the path directly from `opts[:views]`. I think I've seen the mechanics at work though, looking at how `rodauth` for example provides templates for the views.

**Comment by jeremyevans** [comment]
Roda doesn't ship with direct support for what you want, but something like this may work for you (untested):

```ruby
def template_path(opts)
  path = super
  paths = PATHS.each
  until File.file?(path)
    path = super(opts.merge(views: paths.next))
  end
  path
end
```

**Comment by muellerj** [comment]
Works like a charm - thank you very much :+1: 

I might go one step further and (mis-)use the `render_opts[:allowed_paths]` to iterate over? Was the original intention to limit misuse when someone sets the `path` dynamically?

**Comment by jeremyevans** [comment]
Correct, `:allowed_paths` is designed as a security check, to avoid RCE vulnerabilities in case a developer accidentally uses user-provided values when constructing paths to render.  Iterating over `:allowed_paths` shouldn't be a problem

**Comment by muellerj** [comment]
Awesome - thank you again!

I'll leave this here in case someone else stumbles across a similar requirement:

``` ruby
  def template_path(opts)
    render_opts[:allowed_paths]
      .lazy
      .map  { |path| super(opts.merge(views: path)) }
      .find { |path| File.file?(path) }
      .tap  { |path| raise "Could not find template #{opts[:template]} in " +
              "#{render_opts[:allowed_paths].join(":")}" if path.nil? }
  end
```

**Comment by hmdne** [comment]
I have a branch of Roda that wasn't accepted into the upstream (see: #195):

https://github.com/hmdne/roda/commits/tpl_override

As of now, I vendor in a version of those files in my application.
